File: iosApp/iosApp/ViewModels/StoryDetailViewModel.swift
import Foundation
import SharedLogic

@MainActor
public final class StoryDetailViewModel: ObservableObject {
    @Published public private(set) var story: StoryDetail? = nil
    @Published public private(set) var chapters: [Chapter] = []
    @Published public private(set) var isInLibrary: Bool = false
    @Published public private(set) var isDownloaded: Bool = false
    @Published public private(set) var isDownloading: Bool = false
    @Published public private(set) var downloadProgress: Float = 0
    @Published public private(set) var readingProgress: ReadingProgress? = nil
    @Published public private(set) var isLoading: Bool = false
    @Published public private(set) var errorMessage: String? = nil
    @Published public var downloadErrorMessage: String? = nil
    @Published public var downloadSuccessMessage: String? = nil

    private let storySourceRegistry: StorySourceRegistry
    private let readerRepository: ReaderRepository
    private let localCache: LocalStoryCache
    private let downloadManager: DownloadManager

    public init(
        storySourceRegistry: StorySourceRegistry = KoinHelper().storySourceRegistry,
        readerRepository: ReaderRepository = KoinHelper().readerRepository,
        localCache: LocalStoryCache = KoinHelper().localStoryCache,
        downloadManager: DownloadManager = KoinHelper().downloadManager
    ) {
        self.storySourceRegistry = storySourceRegistry
        self.readerRepository = readerRepository
        self.localCache = localCache
        self.downloadManager = downloadManager
    }

    public func loadStory(storyId: String) {
        let parsedStoryId = StoryId.companion.from(compositeValue: storyId)
        let source = storySourceRegistry.getSource(sourceId: parsedStoryId.sourceId) ?? storySourceRegistry.getDefaultSource()
        guard let source = source else {
            self.errorMessage = "No active story source."
            return
        }

        isLoading = true
        errorMessage = nil

        Task {
            do {
                self.isDownloaded = (try await downloadManager.isStoryDownloaded(storyId: parsedStoryId)).boolValue

                let detailResult = try await source.getStoryDetail(rawId: parsedStoryId.rawId)
                if let detail = detailResult.getOrNull() as? StoryDetail {
                    self.story = detail
                    try? await localCache.cacheStory(story: detail.story, accessedAt: Int64(Date().timeIntervalSince1970 * 1000))
                } else if let err = detailResult.errorOrNull() {
                    self.errorMessage = err.message
                }

                // Check cached chapters
                let cachedChapters = (try? await localCache.getCachedChapters(storyId: parsedStoryId)) ?? []
                if !cachedChapters.isEmpty {
                    self.chapters = cachedChapters
                } else {
                    let chaptersResult = try await source.getChapterList(rawId: parsedStoryId.rawId)
                    if let chapterList = chaptersResult.getOrNull() as? [Chapter] {
                        self.chapters = chapterList
                        try? await localCache.cacheChapters(chapters: chapterList)
                    }
                }

                self.isLoading = false
            } catch {
                self.errorMessage = error.localizedDescription
                self.isLoading = false
            }
        }
    }

    public func startDownload() {
        guard let storyDetail = story, let downloadUrl = chapters.first?.url else { return }

        isDownloading = true
        downloadProgress = 0.05
        downloadErrorMessage = nil
        downloadSuccessMessage = nil

        Task {
            do {
                _ = try await downloadManager.startDownload(story: storyDetail.story, downloadUrl: downloadUrl) { progress in
                    Task { @MainActor in
                        self.downloadProgress = progress.progress
                        if progress.status == .completed {
                            self.isDownloading = false
                            self.isDownloaded = true
                            self.isInLibrary = true
                            self.downloadSuccessMessage = "Đã tải xong toàn bộ tác phẩm!"
                            let parsedStoryId = StoryId.companion.from(compositeValue: storyDetail.story.id.value)
                            if let reloaded = try? await self.localCache.getCachedChapters(storyId: parsedStoryId) {
                                self.chapters = reloaded
                            }
                        } else if progress.status == .failed {
                            self.isDownloading = false
                            self.downloadErrorMessage = progress.errorMessage ?? "Tải truyện thất bại. Vui lòng thử lại."
                        }
                    }
                }
            } catch {
                Task { @MainActor in
                    self.isDownloading = false
                    self.downloadErrorMessage = error.localizedDescription
                }
            }
        }
    }

    public func toggleLibrary() {
        self.isInLibrary.toggle()
    }
}

File: androidApp/src/main/kotlin/com/slowbuild/storyverse/storyverse/ui/detail/StoryDetailViewModel.kt
package com.slowbuild.storyverse.storyverse.ui.detail

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.slowbuild.storyverse.core.result.AppResult
import com.slowbuild.storyverse.data.local.LocalStoryCache
import com.slowbuild.storyverse.data.local.room.dao.StoryDao
import com.slowbuild.storyverse.data.local.room.entity.StoryEntity
import com.slowbuild.storyverse.domain.download.DownloadManager
import com.slowbuild.storyverse.domain.download.DownloadProgress
import com.slowbuild.storyverse.domain.model.Chapter
import com.slowbuild.storyverse.domain.model.DownloadStatus
import com.slowbuild.storyverse.domain.model.ReadingProgress
import com.slowbuild.storyverse.domain.model.Story
import com.slowbuild.storyverse.domain.model.StoryDetail
import com.slowbuild.storyverse.domain.model.StoryId
import com.slowbuild.storyverse.domain.repository.ReaderRepository
import com.slowbuild.storyverse.domain.source.StorySourceRegistry
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

data class StoryDetailUiState(
    val isLoading: Boolean = true,
    val story: Story? = null,
    val chapters: List<Chapter> = emptyList(),
    val readingProgress: ReadingProgress? = null,
    val downloadProgress: DownloadProgress? = null,
    val isDownloaded: Boolean = false,
    val isInLibrary: Boolean = false,
    val errorMessage: String? = null
)

class StoryDetailViewModel(
    private val storySourceRegistry: StorySourceRegistry,
    private val localStoryCache: LocalStoryCache,
    private val storyDao: StoryDao,
    private val readerRepository: ReaderRepository,
    private val downloadManager: DownloadManager
) : ViewModel() {

    private val _uiState = MutableStateFlow(StoryDetailUiState())
    val uiState: StateFlow<StoryDetailUiState> = _uiState.asStateFlow()

    private var currentStoryId: StoryId? = null

    fun loadStory(rawOrCompositeId: String) {
        val storyId = StoryId.from(rawOrCompositeId)
        currentStoryId = storyId

        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true, errorMessage = null) }

            // Observe reading progress
            launch {
                readerRepository.observeReadingProgress(storyId).collect { progress ->
                    _uiState.update { it.copy(readingProgress = progress) }
                }
            }

            // Observe download progress
            launch {
                downloadManager.observeDownloadProgress(storyId).collect { progress ->
                    _uiState.update {
                        it.copy(
                            downloadProgress = progress,
                            isDownloaded = progress?.status == DownloadStatus.COMPLETED || it.isDownloaded
                        )
                    }
                    if (progress?.status == DownloadStatus.COMPLETED) {
                        val refreshedChapters = localStoryCache.getCachedChapters(storyId)
                        if (refreshedChapters.isNotEmpty()) {
                            _uiState.update { it.copy(chapters = refreshedChapters, isDownloaded = true) }
                        }
                    }
                }
            }

            // Check if already downloaded
            val isDownloaded = downloadManager.isStoryDownloaded(storyId)

            // Check cached story first
            val cachedStory = localStoryCache.getCachedStory(storyId)
            val cachedChapters = localStoryCache.getCachedChapters(storyId)

            if (cachedStory != null && cachedChapters.isNotEmpty()) {
                _uiState.update {
                    it.copy(
                        isLoading = false,
                        story = cachedStory,
                        chapters = cachedChapters,
                        isDownloaded = isDownloaded,
                        isInLibrary = cachedStory.inLibrary
                    )
                }
            }

            // Fetch latest from remote source
            val source = storySourceRegistry.getSource(storyId.sourceId)
                ?: storySourceRegistry.getDefaultSource()

            if (source != null) {
                when (val result = source.getStoryDetail(storyId.rawId)) {
                    is AppResult.Success -> {
                        val detail = result.data
                        localStoryCache.cacheStory(detail.story)
                        if (cachedChapters.isEmpty()) {
                            localStoryCache.cacheChapters(detail.chapters)
                        }

                        val latestChapters = localStoryCache.getCachedChapters(storyId).ifEmpty { detail.chapters }

                        _uiState.update {
                            it.copy(
                                isLoading = false,
                                story = detail.story,
                                chapters = latestChapters,
                                isDownloaded = isDownloaded,
                                isInLibrary = cachedStory?.inLibrary ?: detail.story.inLibrary,
                                errorMessage = null
                            )
                        }
                    }
                    is AppResult.Error -> {
                        if (_uiState.value.story == null) {
                            _uiState.update {
                                it.copy(
                                    isLoading = false,
                                    errorMessage = result.error.message
                                )
                            }
                        }
                    }
                }
            } else if (_uiState.value.story == null) {
                _uiState.update {
                    it.copy(
                        isLoading = false,
                        errorMessage = "Không tìm thấy nguồn truyện"
                    )
                }
            }
        }
    }

    fun startDownload() {
        val story = _uiState.value.story ?: return
        val downloadUrl = _uiState.value.chapters.firstOrNull()?.url ?: return

        viewModelScope.launch {
            downloadManager.downloadStory(story, downloadUrl).collect { progress ->
                _uiState.update { it.copy(downloadProgress = progress) }
                if (progress.status == DownloadStatus.COMPLETED) {
                    val updatedChapters = localStoryCache.getCachedChapters(story.id)
                    _uiState.update {
                        it.copy(
                            chapters = updatedChapters,
                            isDownloaded = true,
                            isInLibrary = true
                        )
                    }
                }
            }
        }
    }

    fun toggleLibrary() {
        val story = _uiState.value.story ?: return
        val nextInLibrary = !_uiState.value.isInLibrary
        _uiState.update { it.copy(isInLibrary = nextInLibrary) }

        viewModelScope.launch {
            val updatedStory = story.copy(inLibrary = nextInLibrary)
            storyDao.insertOrUpdate(StoryEntity.fromDomain(updatedStory, lastAccessedAt = System.currentTimeMillis()))
        }
    }
}
